src/neptune_query/internal/retrieval/global_search.py

```diff
@@ -0,0 +1,194 @@
+) -> Generator[util.Page[GlobalRunSearchEntry], None, None]:
+    batch_size = env.NEPTUNE_QUERY_SYS_ATTRS_BATCH_SIZE.get()
+
+    yield from util.fetch_pages(
+        client=client,
+        fetch_page=_fetch_entries_page,
+        process_page=ft.partial(_process_entries_page, container_type=container_type),
+        make_new_page_params=ft.partial(_make_next_page_params, limit=limit, batch_size=batch_size),
+        initial_params=GlobalSearchParamsDTO(
+            experiment_leader=container_type == ContainerType.EXPERIMENT,
+            pagination=QueryLeaderboardParamsPaginationDTO(
+                limit=min(batch_size, limit) if limit is not None else batch_size,
+                offset=0,
+            ),
+            query=NqlQueryParamsDTO(query=filter_.to_query()) if filter_ is not None else UNSET,
+            sorting=QueryLeaderboardParamsSortingParamsDTO(
+                sort_by=QueryLeaderboardParamsFieldDTO(
+                    name=sort_by.name,
+                    type=AttributeTypeDTO(
+                        map_attribute_type_python_to_backend(sort_by.type) if sort_by.type is not None else "string"
+                    ),
+                    aggregation_mode=QueryLeaderboardParamsFieldDTOAggregationMode(sort_by.aggregation)
+                    if sort_by.aggregation is not None
+                    else UNSET,
+                ),
+                dir_=QueryLeaderboardParamsSortingParamsDTODir.ASCENDING
+                if sort_direction == "asc"
+                else QueryLeaderboardParamsSortingParamsDTODir.DESCENDING,
+            ),
+        ),
+    )
+
+
+def _fetch_entries_page(
+    client: AuthenticatedClient,
+    params: GlobalSearchParamsDTO,
+) -> ProtoLeaderboardEntriesSearchResultDTO:
+    logger.debug(f"Calling search_l_global_entries_proto with params: {params}")
+
+    call_api = retry.handle_errors_default(with_neptune_client_metadata(search_l_global_entries_proto.sync_detailed))
+    response = call_api(client=client, body=params)
+
+    logger.debug(
+        f"search_l_global_entries_proto response status: {response.status_code}, "
+        f"content length: {len(response.content) if response.content else 'no content'}"
+    )
+    return ProtoLeaderboardEntriesSearchResultDTO.FromString(response.content)
+
+
+def _make_next_page_params(
+    current_params: GlobalSearchParamsDTO,
+    data: Optional[ProtoLeaderboardEntriesSearchResultDTO],
+    *,
+    limit: Optional[int],
+    batch_size: int,
+) -> Optional[GlobalSearchParamsDTO]:
+    if data is None:
+        return current_params
+
+    # If the server returned fewer entries than we process in a single request, we've reached the end.
+    if len(data.entries) < batch_size:
+        return None
+
+    return attrs.evolve(
+        current_params,
+        pagination=attrs.evolve(
+            current_params.pagination,
+            limit=min(limit - current_params.pagination.offset, batch_size) if limit is not None else UNSET,
```

Comment:
`min(limit - current_params.pagination.offset, batch_size) `

I think this should be

`min(limit - NEXT_params.pagination.offset, batch_size) `